packages/toolpad-app/src/toolpad/AppEditor/PageEditor/EditorCanvasHost.tsx

```diff
@@ -145,10 +148,40 @@ export default React.forwardRef<EditorCanvasHostHandle, EditorCanvasHostProps>(
 
     const handleRuntimeEvent = useEvent(onRuntimeEvent);
 
+    const iframeKeyDownHandler = React.useCallback(
+      (iframeDocument: Document) => {
+        return (event: KeyboardEvent) => {
+          if (hasFieldFocus(iframeDocument)) {
+            return;
+          }
+
+          const { code, metaKey, shiftKey } = event;
+          const undoShortcut = code === 'KeyZ' && metaKey;
+          const redoShortcut = undoShortcut && shiftKey;
+
+          if (redoShortcut) {
+            domApi.redo();
+          } else if (undoShortcut) {
+            domApi.undo();
+          }
+        };
+      },
+      [domApi],
+    );
+
     const handleFrameLoad = React.useCallback(() => {
       invariant(frameRef.current, 'Iframe ref not attached');
-      setContentWindow(frameRef.current.contentWindow);
-    }, []);
+
+      const iframeWindow = frameRef.current.contentWindow;
+      setContentWindow(iframeWindow);
+
+      const keyDownHandler = iframeKeyDownHandler(iframeWindow.document);
+
+      iframeWindow?.addEventListener('keydown', keyDownHandler);
```

Comment:
I've tried getting `useShortcut` to work with this iframe flow, but seems because it's loaded async hook just didn't pick up either I just couldn't figure out what's the proper way. I'm open to suggestions if this can be improved.

Reply:
Just an idea, but what if we just add the undo/redo buttons as per design and leave the shortcut as a follow up? It would allow us to concentrate on functionality first without all the shortcuts clutter.

Reply:
I can add buttons, but why would we want to omit key shortcut? IMO it's way more comfortable to use shortcuts than to every time go back and forth to the buttons 🤔